File: scripts/Reconstruct Original Digits from English/Reconstruct Original Digits from English.js
/**
 * @param {string} s
 * @return {string}
 */
 
var originalDigits = function(s) {
    const numberToWord = {
        0: 'zero',
        1: 'one',
        2: 'two',
        3: 'three',
        4: 'four',
        5: 'five',
        6: 'six',
        7: 'seven',
        8: 'eight',
        9: 'nine'
    }
    
    const letterToDigits = {}
    
    for(const key in numberToWord) {
        const currWord = numberToWord[key]
        
        for(const letter of currWord) {
            if(!(letter in letterToDigits)) letterToDigits[letter] = new Set()
            letterToDigits[letter].add(key)
        }
    }
    
    const inputFreqs = {}
    
    for(let i = 0; i < s.length; i++) {
        const currChar = s[i]
        
        if(!(currChar in inputFreqs)) inputFreqs[currChar] = 0
        inputFreqs[currChar]++
    }
    
    const letters = Object.keys(inputFreqs)
    
    const res = dfs(letters[0])
    
    return [...res].sort().join('')
    
    function dfs(currLetter) {
        if(!isValid(inputFreqs)) return null
        if(getTotalRemaining(inputFreqs) === 0) return []
        const possibleDigits = letterToDigits[currLetter]
        
        for(const digit of [...possibleDigits]) {
            const wordRepresentation = numberToWord[digit]
            
            subtract(wordRepresentation)
            
            if(!isValid(inputFreqs)) {
                addBack(wordRepresentation)
                continue
            }
            
            const nextLetter = getNext(inputFreqs)
            const nextDigits = dfs(nextLetter)
            
            if(nextDigits !== null) return [digit] + nextDigits
            
            addBack(wordRepresentation)
        }
        
        return null
    }
    
    function isValid(inputFreqs) {
        for(const key in inputFreqs) {
            const count = inputFreqs[key]
            
            if(count < 0) return false
        }
        
        return true
    }
    
    function getTotalRemaining(inputFreqs) {
        let sum = 0
        
        for(const key in inputFreqs) {
            const count = inputFreqs[key]
            
            sum += count
        }
        
        
        return sum
    }
    
    function subtract(word) {
        for(const char of word) {
            if(!(char in inputFreqs)) inputFreqs[char] = 0
            inputFreqs[char]--
        }
    }
    
    
    function addBack(word) {
        for(const char of word) {
            inputFreqs[char]++
        }
    }
    
    function getNext(inputFreqs) {
        for(const key in inputFreqs) {
            const count = inputFreqs[key]
            
            if(count > 0) return key
        }
        
        return null
    }
};

File: scripts/Reconstruct Original Digits from English/Reconstruct Original Digits from English.cpp
class Solution {
public:
    
    vector<string> digits{"zero","one","two","three","four","five","six","seven","eight","nine"};
    void fun(int x,string &ans,vector<int> &m){
        for(int i = x;i<=9;i+=2){
            string t = digits[i];
            if(t.length() == 3){
                if(m[t[0]-'a'] > 0 && m[t[1]-'a'] > 0 && m[t[2]-'a'] > 0){
                    int min_o = min({m[t[0]-'a'] ,m[t[1]-'a'] , m[t[2]-'a']});
                    m[t[0]-'a'] -= min_o;
                    m[t[1]-'a'] -= min_o;
                    m[t[2]-'a'] -= min_o;
                    while(min_o--) ans += (char)(i +'0');
                }
            }else if(t.length() == 4){
                if(m[t[0]-'a'] > 0 && m[t[1]-'a'] > 0 && m[t[2]-'a'] > 0 && m[t[3]-'a'] > 0){
                    int min_o = min({m[t[0]-'a'] ,m[t[1]-'a'] , m[t[2]-'a'], m[t[3]-'a']});
                    m[t[0]-'a'] -= min_o;
                    m[t[1]-'a'] -= min_o;
                    m[t[2]-'a'] -= min_o;
                    m[t[3]-'a'] -= min_o;
                    while(min_o--) ans += (char)(i +'0');
                }
            }else if(t.length() == 5){
                if(m[t[0]-'a'] > 0 && m[t[1]-'a'] > 0 && m[t[2]-'a'] > 0 && m[t[3]-'a'] > 0 && m[t[4]-'a'] > 0){
                    int min_o = min({m[t[0]-'a'] ,m[t[1]-'a'] , m[t[2]-'a'], m[t[3]-'a'], m[t[4]-'a']});
                    m[t[0]-'a'] -= min_o;
                    m[t[1]-'a'] -= min_o;
                    m[t[2]-'a'] -= min_o;
                    m[t[3]-'a'] -= min_o;
                    m[t[4]-'a'] -= min_o;
                    while(min_o--) ans += (char)(i +'0');
                }
            }
        }
    }
    
    string originalDigits(string s) {
        string ans;
        vector<int> m(26,0);
        for(auto x:s) m[x-'a']++;
        fun(0,ans,m);
        fun(1,ans,m);
        sort(ans.begin(),ans.end());
        return ans;
    }
};
